### PR TITLE
cryptutil: add envelope encryption w/key encryption key and data encryption key

### PR DESCRIPTION
**Pull request by calebdoxsey**
## Summary
Add `KeyEncryptionKey` and `DataEncryptionKey` for performing envelope encryption.

## Checklist
- [ ] reference any related issues
- [ ] updated docs
- [x] updated unit tests
- [ ] updated UPGRADING.md
- [x] add appropriate tag (`improvement` / `bug` / etc)
- [x] ready for review

### CODE REVIEW

File: pkg/cryptutil/dek.go [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+			DataEncryptionKeySize, len(raw))
+	}
+	dek := new(DataEncryptionKey)
+	copy(dek.data[:], raw)
+	dek.cipher, _ = chacha20poly1305.NewX(raw) // only errors on invalid size
+	return dek, nil
+}
+
+// GenerateDataEncryptionKey generates a new random data encryption key.
+func GenerateDataEncryptionKey() (*DataEncryptionKey, error) {
```
Comment (desimone) [inline pkg/cryptutil/dek.go:41]: I suppose we could use the existing function `NewKey` which I think does the same thing?

Comment (desimone) [reply pkg/cryptutil/dek.go:41]: or, `randomBytes(c)`

File: pkg/cryptutil/dek.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+}
+
+// EncryptString encrypts a string using the data encryption key and base64 encoding.
+func (dek *DataEncryptionKey) EncryptString(plaintext string) string {
+	bs := dek.Encrypt([]byte(plaintext))
+	return base64.StdEncoding.EncodeToString(bs)
+}
+
+// KeyBytes returns the private key encryption key's raw bytes.
+func (dek *DataEncryptionKey) KeyBytes() []byte {
```
Comment (desimone) [inline pkg/cryptutil/dek.go:80]: Could we add tests for this func?

File: pkg/cryptutil/dek.go [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+	return string(plaintextBytes), nil
+}
+
+// Encrypt encrypts data using the data encryption key.
+func (dek *DataEncryptionKey) Encrypt(plaintext []byte) []byte {
+	return Encrypt(dek.cipher, plaintext, nil)
+}
+
+// EncryptString encrypts a string using the data encryption key and base64 encoding.
+func (dek *DataEncryptionKey) EncryptString(plaintext string) string {
```
Comment (desimone) [inline pkg/cryptutil/dek.go:74]: Could we add tests for this func?

File: pkg/cryptutil/dek.go [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+	return NewDataEncryptionKey(raw)
+}
+
+// Decrypt decrypts encrypted data using the data encryption key.
+func (dek *DataEncryptionKey) Decrypt(ciphertext []byte) ([]byte, error) {
+	return Decrypt(dek.cipher, ciphertext, nil)
+}
+
+// DecryptString decrypts an encrypted string using the data encryption key and base64 encoding.
+func (dek *DataEncryptionKey) DecryptString(ciphertext string) (string, error) {
```
Comment (desimone) [inline pkg/cryptutil/dek.go:56]: Could we add tests for this func?

File: pkg/cryptutil/kek.go [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+	}
+	kek := new(PrivateKeyEncryptionKey)
+	kek.id = id
+	copy(kek.data[:], raw)
+	return kek, nil
+}
+
+// GenerateKeyEncryptionKey generates a new random key encryption key.
+func GenerateKeyEncryptionKey() (*PrivateKeyEncryptionKey, error) {
+	raw := make([]byte, KeyEncryptionKeySize)
```
Comment (desimone) [inline pkg/cryptutil/kek.go:60]: can we reuse existing `randomBytes(c)` func?

File: pkg/cryptutil/kek.go [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+	_, err := io.ReadFull(rand.Reader, raw)
+	if err != nil {
+		return nil, fmt.Errorf("cryptutil: error generating random data encryption key: %w", err)
+	}
+	return NewPrivateKeyEncryptionKey(uuid.New().String(), raw)
```
Comment (desimone) [inline pkg/cryptutil/kek.go:65]: Thoughts on using the sha256-hmac of the key itself for this value to:
1. Limit depedencies
2. If, for whatever reason, you have just the private key, you can find out the associated key-id

File: pkg/cryptutil/kek.go (hunk 5 above)
Comment (desimone) [reply pkg/cryptutil/kek.go:60]: (the downside being that randomBytes does panic on lack of entropy, but I would argue that is one of the few times panic'ing makes sense.

File: pkg/cryptutil/kek.go (hunk 6 above)
Comment (calebdoxsey) [reply pkg/cryptutil/kek.go]: I used our existing `Hash` function and base58.

**Review by desimone — APPROVED** [review]

**Pull request merged**